Source/Classes/PINSpeedRecorder.m

```diff
@@ -66,69 +77,113 @@ - (instancetype)init
     return self;
 }
 
-- (void)addTaskBPS:(float)bytesPerSecond endDate:(NSDate *)endDate
+- (void)processMetrics:(NSURLSessionTaskMetrics *)metrics forTask:(NSURLSessionTask *)task host:(NSString *)host
 {
-    //if bytesPerSecond is less than or equal to zero, ignore.
-    if (bytesPerSecond <= 0) {
+    NSDate *requestStart = [NSDate distantFuture];
+    NSDate *firstByte = [NSDate distantFuture];
+    NSDate *requestEnd = [NSDate distantPast];
+    int64_t contentLength = task.countOfBytesReceived;
+    
+    for (NSURLSessionTaskTransactionMetrics *metric in metrics.transactionMetrics) {
+        if (metric.requestStartDate == nil || metric.responseStartDate == nil) {
+            //Only evaluate requests which completed their first byte.
+            return;
+        }
+        if ([requestStart compare:metric.requestStartDate] != NSOrderedAscending) {
+            requestStart = metric.requestStartDate;
+        }
+        if ([firstByte compare:metric.responseStartDate] != NSOrderedAscending) {
+            firstByte = metric.responseStartDate;
+        }
+        if ([requestEnd compare:metric.responseEndDate] != NSOrderedDescending) {
+            requestEnd = metric.responseEndDate;
+        }
+    }
+    
+    if ([requestStart isEqual:[NSDate distantFuture]] || [firstByte isEqual:[NSDate distantFuture]] || [requestEnd isEqual:[NSDate distantPast]] || contentLength == 0) {
         return;
     }
+
+    PINSpeedMeasurement *measurement = [[PINSpeedMeasurement alloc] init];
+    measurement.count = 1;
+    measurement.bytesPerSecond = contentLength / [requestEnd timeIntervalSinceDate:requestStart];
+    measurement.startAdjustedBytesPerSecond = contentLength / [requestEnd timeIntervalSinceDate:firstByte];
+    measurement.timeToFirstByte = [firstByte timeIntervalSinceDate:requestStart];
     
+    [self updateSpeedsForHost:host
+               bytesPerSecond:contentLength / [requestEnd timeIntervalSinceDate:requestStart]
+  startAdjustedBytesPerSecond:contentLength / [requestEnd timeIntervalSinceDate:firstByte]
+              timeToFirstByte:[firstByte timeIntervalSinceDate:requestStart]];
```

Comment:
To reduce the amount of math in here, maybe just pass `bytesPerSecond:measurement.bytesPerSecond` etc?

Reply:
Whoops, that measurement object was supposed to be deleted :)